# Proyek Analisis Data: E-Commerce Public dataset
- Nama: WIFQO AROVA SYAMS
- Email: wifqoarova17@gmail.com
- Id Dicoding: WIFQO AROVA SYAMS

## Menentukan Pertanyaan Bisnis

- pertanyaan 1
- pertanyaan 2

## Menyiapkan semua library yang dibutuhkan

In [2]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

## Data Wrangling

### Gathering Data

In [3]:
# load all datasets from "E-Commerce Public Dataset" folder, then instanciate it as dataframe

customers_df = pd.read_csv("E-Commerce Public Dataset/customers_dataset.csv")
sellers_df = pd.read_csv("E-Commerce Public Dataset/sellers_dataset.csv")
products_df = pd.read_csv("E-Commerce Public Dataset/products_dataset.csv")
geolocation_df = pd.read_csv("E-Commerce Public Dataset/geolocation_dataset.csv")
product_category_name_translation_df = pd.read_csv("E-Commerce Public Dataset/product_category_name_translation.csv")
orders_df = pd.read_csv("E-Commerce Public Dataset/orders_dataset.csv")
order_reviews_df = pd.read_csv("E-Commerce Public Dataset/order_reviews_dataset.csv")
order_items_df = pd.read_csv("E-Commerce Public Dataset/order_items_dataset.csv")
order_payments_df = pd.read_csv("E-Commerce Public Dataset/order_payments_dataset.csv")

### Assessing Data

#### 1. Assessing Customer DataFrame

data assessing result:
- there are don't have a wrong datatype
- there are no missing value
- there are no duplicated value
- there are no irregularities in statistic parameters (no column to be calculated)

In [4]:
# check for datatype
customers_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 99441 entries, 0 to 99440
Data columns (total 5 columns):
 #   Column                    Non-Null Count  Dtype 
---  ------                    --------------  ----- 
 0   customer_id               99441 non-null  object
 1   customer_unique_id        99441 non-null  object
 2   customer_zip_code_prefix  99441 non-null  int64 
 3   customer_city             99441 non-null  object
 4   customer_state            99441 non-null  object
dtypes: int64(1), object(4)
memory usage: 3.8+ MB


In [5]:
# check for missing value
customers_df.isna().sum()

customer_id                 0
customer_unique_id          0
customer_zip_code_prefix    0
customer_city               0
customer_state              0
dtype: int64

In [6]:
# check for duplicated value
print("Duplicated Value:", customers_df.duplicated().sum())

Duplicated Value: 0


In [7]:
# check for statistic parameters
customers_df.describe()

customer_zip_code_prefix
count              99441.000000
mean               35137.474583
std                29797.938996
min                 1003.000000
25%                11347.000000
50%                24416.000000
75%                58900.000000
max                99990.000000

#### 2. Assessing Sellers DataFrame

data assessing result:
- there are don't have a wrong datatype
- there are no missing value
- there are no duplicated value
- there are no irregularities in statistic parameters (no column to be calculated)

In [8]:
# check for datatype
sellers_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3095 entries, 0 to 3094
Data columns (total 4 columns):
 #   Column                  Non-Null Count  Dtype 
---  ------                  --------------  ----- 
 0   seller_id               3095 non-null   object
 1   seller_zip_code_prefix  3095 non-null   int64 
 2   seller_city             3095 non-null   object
 3   seller_state            3095 non-null   object
dtypes: int64(1), object(3)
memory usage: 96.8+ KB


In [9]:
# check for missing value
sellers_df.isna().sum()

seller_id                 0
seller_zip_code_prefix    0
seller_city               0
seller_state              0
dtype: int64

In [10]:
# check for duplicated value
print("Duplicated value:", sellers_df.duplicated().sum())

Duplicated value: 0


In [11]:
# check for parameter statistics
sellers_df.describe()

seller_zip_code_prefix
count             3095.000000
mean             32291.059451
std              32713.453830
min               1001.000000
25%               7093.500000
50%              14940.000000
75%              64552.500000
max              99730.000000

#### 3. Assessing Product DataFrame

data assessing result:
- there are don't have a wrong datatype
- there are have a lot missing value
- there are no duplicated value
- there are no irregularities in statistic parameters (no column to be calculated)

In [12]:
# check for datatype
products_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 32951 entries, 0 to 32950
Data columns (total 9 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   product_id                  32951 non-null  object 
 1   product_category_name       32341 non-null  object 
 2   product_name_lenght         32341 non-null  float64
 3   product_description_lenght  32341 non-null  float64
 4   product_photos_qty          32341 non-null  float64
 5   product_weight_g            32949 non-null  float64
 6   product_length_cm           32949 non-null  float64
 7   product_height_cm           32949 non-null  float64
 8   product_width_cm            32949 non-null  float64
dtypes: float64(7), object(2)
memory usage: 2.3+ MB


In [13]:
# check for missing value
products_df.isna().sum()

product_id                      0
product_category_name         610
product_name_lenght           610
product_description_lenght    610
product_photos_qty            610
product_weight_g                2
product_length_cm               2
product_height_cm               2
product_width_cm                2
dtype: int64

In [14]:
# check for duplicate value
print("Duplicated Value:", products_df.duplicated().sum())

Duplicated Value: 0


In [15]:
# check for statistic parameters
products_df.describe()

product_name_lenght  product_description_lenght  product_photos_qty  \
count         32341.000000                32341.000000        32341.000000   
mean             48.476949                  771.495285            2.188986   
std              10.245741                  635.115225            1.736766   
min               5.000000                    4.000000            1.000000   
25%              42.000000                  339.000000            1.000000   
50%              51.000000                  595.000000            1.000000   
75%              57.000000                  972.000000            3.000000   
max              76.000000                 3992.000000           20.000000   

       product_weight_g  product_length_cm  product_height_cm  \
count      32949.000000       32949.000000       32949.000000   
mean        2276.472488          30.815078          16.937661   
std         4282.038731          16.914458          13.637554   
min            0.000000           7.000000           2.000000   
25%          300.000000          18.000000           8.000000   
50%          700.000000          25.000000          13.000000   
75%         1900.000000          38.000000          21.000000   
max        40425.000000         105.000000         105.000000   

       product_width_cm  
count      32949.000000  
mean          23.196728  
std           12.079047  
min            6.000000  
25%           15.000000  
50%           20.000000  
75%           30.000000  
max          118.000000

#### 4. Assessing Geolocations DataFrame

data assessing result:
- there are don't have a wrong datatype
- there are no missing value
- there are have 261831 duplicated value
- there are no irregularities in statistic parameters (no column to be calculated)

In [16]:
# check for datatype
geolocation_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000163 entries, 0 to 1000162
Data columns (total 5 columns):
 #   Column                       Non-Null Count    Dtype  
---  ------                       --------------    -----  
 0   geolocation_zip_code_prefix  1000163 non-null  int64  
 1   geolocation_lat              1000163 non-null  float64
 2   geolocation_lng              1000163 non-null  float64
 3   geolocation_city             1000163 non-null  object 
 4   geolocation_state            1000163 non-null  object 
dtypes: float64(2), int64(1), object(2)
memory usage: 38.2+ MB


In [17]:
# check for missing value
geolocation_df.isna().sum()

geolocation_zip_code_prefix    0
geolocation_lat                0
geolocation_lng                0
geolocation_city               0
geolocation_state              0
dtype: int64

In [18]:
# check for duplicated data
geolocation_df.duplicated().sum()

261831

In [19]:
# check for statistic parameters
geolocation_df.describe()

geolocation_zip_code_prefix  geolocation_lat  geolocation_lng
count                 1.000163e+06     1.000163e+06     1.000163e+06
mean                  3.657417e+04    -2.117615e+01    -4.639054e+01
std                   3.054934e+04     5.715866e+00     4.269748e+00
min                   1.001000e+03    -3.660537e+01    -1.014668e+02
25%                   1.107500e+04    -2.360355e+01    -4.857317e+01
50%                   2.653000e+04    -2.291938e+01    -4.663788e+01
75%                   6.350400e+04    -1.997962e+01    -4.376771e+01
max                   9.999000e+04     4.506593e+01     1.211054e+02

#### 5. Assessing Product Category Name Translation DataFrame
data assessing result:
- there are don't have a wrong datatype
- there are no missing value
- there are no duplicated value
- there are no irregularities in statistic parameters (no column to be calculated)

In [20]:
# check for datatype
product_category_name_translation_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 71 entries, 0 to 70
Data columns (total 2 columns):
 #   Column                         Non-Null Count  Dtype 
---  ------                         --------------  ----- 
 0   product_category_name          71 non-null     object
 1   product_category_name_english  71 non-null     object
dtypes: object(2)
memory usage: 1.2+ KB


In [21]:
# check for missing value
product_category_name_translation_df.isna().sum()

product_category_name            0
product_category_name_english    0
dtype: int64

In [22]:
# check for duplicated value
product_category_name_translation_df.duplicated().sum()

0

#### 6. Assessing Orders DataFrame

data assessing result:
- there are have a wrong datatype on timestamps columns
- there are have a lot missing value
- there are no duplicated value
- there are no irregularities in statistic parameters (no column to be calculated)

In [23]:
# check for datatype
orders_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 99441 entries, 0 to 99440
Data columns (total 8 columns):
 #   Column                         Non-Null Count  Dtype 
---  ------                         --------------  ----- 
 0   order_id                       99441 non-null  object
 1   customer_id                    99441 non-null  object
 2   order_status                   99441 non-null  object
 3   order_purchase_timestamp       99441 non-null  object
 4   order_approved_at              99281 non-null  object
 5   order_delivered_carrier_date   97658 non-null  object
 6   order_delivered_customer_date  96476 non-null  object
 7   order_estimated_delivery_date  99441 non-null  object
dtypes: object(8)
memory usage: 6.1+ MB


In [24]:
# check for missing value
orders_df.isna().sum()

order_id                            0
customer_id                         0
order_status                        0
order_purchase_timestamp            0
order_approved_at                 160
order_delivered_carrier_date     1783
order_delivered_customer_date    2965
order_estimated_delivery_date       0
dtype: int64

In [25]:
# check for duplicated value
orders_df.duplicated().sum()

0

In [26]:
# check for statistic parameters
orders_df.describe()

order_id                       customer_id  \
count                              99441                             99441   
unique                             99441                             99441   
top     e481f51cbdc54678b7cc49136f2d6af7  9ef432eb6251297304e76186b10a928d   
freq                                   1                                 1   

       order_status order_purchase_timestamp    order_approved_at  \
count         99441                    99441                99281   
unique            8                    98875                90733   
top       delivered      2018-04-11 10:48:14  2018-02-27 04:31:10   
freq          96478                        3                    9   

       order_delivered_carrier_date order_delivered_customer_date  \
count                         97658                         96476   
unique                        81018                         95664   
top             2018-05-09 15:48:00           2018-05-08 23:38:46   
freq                             47                             3   

       order_estimated_delivery_date  
count                          99441  
unique                           459  
top              2017-12-20 00:00:00  
freq                             522

#### 7. Assessing Order Reviews DataFrame

data assessing result:
- there are have a wrong datatype in each datetime format
- there are have a lot missing value
- there are dont have duplicated data
- there are no irregularities in the statistical parameters

In [27]:
# check for datatype
order_reviews_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 99224 entries, 0 to 99223
Data columns (total 7 columns):
 #   Column                   Non-Null Count  Dtype 
---  ------                   --------------  ----- 
 0   review_id                99224 non-null  object
 1   order_id                 99224 non-null  object
 2   review_score             99224 non-null  int64 
 3   review_comment_title     11568 non-null  object
 4   review_comment_message   40977 non-null  object
 5   review_creation_date     99224 non-null  object
 6   review_answer_timestamp  99224 non-null  object
dtypes: int64(1), object(6)
memory usage: 5.3+ MB


In [28]:
# check for missing value
order_reviews_df.isna().sum()

review_id                      0
order_id                       0
review_score                   0
review_comment_title       87656
review_comment_message     58247
review_creation_date           0
review_answer_timestamp        0
dtype: int64

In [29]:
# check for duplicated value
order_reviews_df.duplicated().sum()

0

In [30]:
# check for statistics parameters
order_reviews_df.describe()

review_score
count  99224.000000
mean       4.086421
std        1.347579
min        1.000000
25%        4.000000
50%        5.000000
75%        5.000000
max        5.000000

#### 8. Assessing Order Items DataFrame

data assessing result:
- there are have wrong datatype in date column
- there are no missing values
- there are no duplicated data
- there are no irregularities in statistic parameters

In [31]:
# check for datatype
order_items_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 112650 entries, 0 to 112649
Data columns (total 7 columns):
 #   Column               Non-Null Count   Dtype  
---  ------               --------------   -----  
 0   order_id             112650 non-null  object 
 1   order_item_id        112650 non-null  int64  
 2   product_id           112650 non-null  object 
 3   seller_id            112650 non-null  object 
 4   shipping_limit_date  112650 non-null  object 
 5   price                112650 non-null  float64
 6   freight_value        112650 non-null  float64
dtypes: float64(2), int64(1), object(4)
memory usage: 6.0+ MB


In [32]:
# check for missing value
order_items_df.isna().sum()

order_id               0
order_item_id          0
product_id             0
seller_id              0
shipping_limit_date    0
price                  0
freight_value          0
dtype: int64

In [33]:
# check duplicated data
order_items_df.duplicated().sum()

0

In [34]:
# check for statictic parameters
order_items_df.describe()

order_item_id          price  freight_value
count  112650.000000  112650.000000  112650.000000
mean        1.197834     120.653739      19.990320
std         0.705124     183.633928      15.806405
min         1.000000       0.850000       0.000000
25%         1.000000      39.900000      13.080000
50%         1.000000      74.990000      16.260000
75%         1.000000     134.900000      21.150000
max        21.000000    6735.000000     409.680000

#### 9. Assessing Order Payments DataFrame

data assessing result:
- there are don't have a wrong datatype
- there are no missing value
- there are no duplicated value
- there are no irregularities in statistic parameters

In [35]:
# check for datatype
order_payments_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 103886 entries, 0 to 103885
Data columns (total 5 columns):
 #   Column                Non-Null Count   Dtype  
---  ------                --------------   -----  
 0   order_id              103886 non-null  object 
 1   payment_sequential    103886 non-null  int64  
 2   payment_type          103886 non-null  object 
 3   payment_installments  103886 non-null  int64  
 4   payment_value         103886 non-null  float64
dtypes: float64(1), int64(2), object(2)
memory usage: 4.0+ MB


In [36]:
# check missing values
order_payments_df.isna().sum()

order_id                0
payment_sequential      0
payment_type            0
payment_installments    0
payment_value           0
dtype: int64

In [37]:
# check duplicated data
order_payments_df.duplicated().sum()

0

In [38]:
# check statistic parameters
order_payments_df.describe()

payment_sequential  payment_installments  payment_value
count       103886.000000         103886.000000  103886.000000
mean             1.092679              2.853349     154.100380
std              0.706584              2.687051     217.494064
min              1.000000              0.000000       0.000000
25%              1.000000              1.000000      56.790000
50%              1.000000              1.000000     100.000000
75%              1.000000              4.000000     171.837500
max             29.000000             24.000000   13664.080000

### Cleaning Data

#### Dataframe list to be cleaning

1. Product DataFrame
    - there are have a lot missing value
2. Geolocation DataFrame
    - there are have 261831 duplicated value
3. Orders DataFrame
    - there are have a wrong datatype on timestamps columns
    - there are have a lot missing value
4. Order Reviews DataFrame
    - there are have a wrong datatype in each datetime format
    - there are have a lot missing value
5. Order Items DataFrame
    - there are have wrong datatype in date column

#### 1. Cleaning Product DataFrame

In [39]:
# check the missing value
products_df.isna().sum()

product_id                      0
product_category_name         610
product_name_lenght           610
product_description_lenght    610
product_photos_qty            610
product_weight_g                2
product_length_cm               2
product_height_cm               2
product_width_cm                2
dtype: int64

In [40]:
# fetch all data that have a missing value based on product_category_name
products_df[products_df.product_category_name.isna()]

product_id product_category_name  \
105    a41e356c76fab66334f36de622ecbd3a                   NaN   
128    d8dee61c2034d6d075997acef1870e9b                   NaN   
145    56139431d72cd51f19eb9f7dae4d1617                   NaN   
154    46b48281eb6d663ced748f324108c733                   NaN   
197    5fb61f482620cb672f5e586bb132eae9                   NaN   
...                                 ...                   ...   
32515  b0a0c5dd78e644373b199380612c350a                   NaN   
32589  10dbe0fbaa2c505123c17fdc34a63c56                   NaN   
32616  bd2ada37b58ae94cc838b9c0569fecd8                   NaN   
32772  fa51e914046aab32764c41356b9d4ea4                   NaN   
32852  c4ceee876c82b8328e9c293fa0e1989b                   NaN   

       product_name_lenght  product_description_lenght  product_photos_qty  \
105                    NaN                         NaN                 NaN   
128                    NaN                         NaN                 NaN   
145                    NaN                         NaN                 NaN   
154                    NaN                         NaN                 NaN   
197                    NaN                         NaN                 NaN   
...                    ...                         ...                 ...   
32515                  NaN                         NaN                 NaN   
32589                  NaN                         NaN                 NaN   
32616                  NaN                         NaN                 NaN   
32772                  NaN                         NaN                 NaN   
32852                  NaN                         NaN                 NaN   

       product_weight_g  product_length_cm  product_height_cm  \
105               650.0               17.0               14.0   
128               300.0               16.0                7.0   
145               200.0               20.0               20.0   
154             18500.0               41.0               30.0   
197               300.0               35.0                7.0   
...                 ...                ...                ...   
32515            1800.0               30.0               20.0   
32589             800.0               30.0               10.0   
32616             200.0               21.0                8.0   
32772            1300.0               45.0               16.0   
32852             700.0               28.0                3.0   

       product_width_cm  
105                12.0  
128                20.0  
145                20.0  
154                41.0  
197                12.0  
...                 ...  
32515              70.0  
32589              23.0  
32616              16.0  
32772              45.0  
32852              43.0  

[610 rows x 9 columns]

In [41]:
# fetch all data that have a missing value based on product_weight_g
products_df[products_df.product_weight_g.isna()]

product_id product_category_name  \
8578   09ff539a621711667c43eba6a3bd8466                 bebes   
18851  5eb564652db742ff8f28759cd8d2652a                   NaN   

       product_name_lenght  product_description_lenght  product_photos_qty  \
8578                  60.0                       865.0                 3.0   
18851                  NaN                         NaN                 NaN   

       product_weight_g  product_length_cm  product_height_cm  \
8578                NaN                NaN                NaN   
18851               NaN                NaN                NaN   

       product_width_cm  
8578                NaN  
18851               NaN

In [42]:
"""
we can see that in one row data there are have a lot missing value in columns,
so better to delete all the data because we have experience lost that data
and the data is useless now because it's not provide the rellevant data
"""
products_df.dropna(axis=0, inplace=True)

#### 2. Cleaning Geolocation DataFrame

In [43]:
# delete all duplicated data
geolocation_df.drop_duplicates(inplace=True)

In [44]:
# check duplicated data
geolocation_df.duplicated().sum()

0

#### 3. Cleaning Order DataFrame

In [45]:
# check before and after cleaned
orders_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 99441 entries, 0 to 99440
Data columns (total 8 columns):
 #   Column                         Non-Null Count  Dtype 
---  ------                         --------------  ----- 
 0   order_id                       99441 non-null  object
 1   customer_id                    99441 non-null  object
 2   order_status                   99441 non-null  object
 3   order_purchase_timestamp       99441 non-null  object
 4   order_approved_at              99281 non-null  object
 5   order_delivered_carrier_date   97658 non-null  object
 6   order_delivered_customer_date  96476 non-null  object
 7   order_estimated_delivery_date  99441 non-null  object
dtypes: object(8)
memory usage: 6.1+ MB


In [46]:
# change date columns to datetime datatype
timestamp_columns = [
    "order_purchase_timestamp", 
    "order_approved_at", 
    "order_delivered_carrier_date", 
    "order_delivered_customer_date", 
    "order_estimated_delivery_date"
]

for column in timestamp_columns:
    orders_df[column] = pd.to_datetime(orders_df[column])

In [47]:
# check the missing value
orders_df.isna().sum()

order_id                            0
customer_id                         0
order_status                        0
order_purchase_timestamp            0
order_approved_at                 160
order_delivered_carrier_date     1783
order_delivered_customer_date    2965
order_estimated_delivery_date       0
dtype: int64

In [48]:
# fetch data that order_approved_at have a missing value
orders_df[orders_df.order_approved_at.isna()]

order_id                       customer_id  \
1130   00b1cb0320190ca0daa2c88b35206009  3532ba38a3fd242259a514ac2b6ae6b6   
1801   ed3efbd3a87bea76c2812c66a0b32219  191984a8ba4cbb2145acb4fe35b69664   
1868   df8282afe61008dc26c6c31011474d02  aa797b187b5466bc6925aaaa4bb3bed1   
2029   8d4c637f1accf7a88a4555f02741e606  b1dd715db389a2077f43174e7a675d07   
2161   7a9d4c7f9b068337875b95465330f2fc  7f71ae48074c0cfec9195f88fcbfac55   
...                                 ...                               ...   
97696  5a00b4d35edffc56b825c3646a99ba9d  6a3bdf004ca96338fb5fad1b8d93c2e6   
98415  227c804e2a44760671a6a5697ea549e4  62e7477e75e542243ee62a0ba73f410f   
98909  e49e7ce1471b4693482d40c2bd3ad196  e4e7ab3f449aeb401f0216f86c2104db   
99283  3a3cddda5a7c27851bd96c3313412840  0b0d6095c5555fe083844281f6b093bb   
99347  a89abace0dcc01eeb267a9660b5ac126  2f0524a7b1b3845a1a57fcf3910c4333   

      order_status order_purchase_timestamp order_approved_at  \
1130      canceled      2018-08-28 15:26:39               NaT   
1801      canceled      2018-09-20 13:54:16               NaT   
1868      canceled      2017-03-04 12:14:30               NaT   
2029      canceled      2018-08-29 16:27:49               NaT   
2161      canceled      2017-05-01 16:12:39               NaT   
...            ...                      ...               ...   
97696     canceled      2017-07-02 15:38:46               NaT   
98415     canceled      2017-09-28 15:02:56               NaT   
98909     canceled      2018-08-07 11:16:28               NaT   
99283     canceled      2018-08-31 16:13:44               NaT   
99347     canceled      2018-09-06 18:45:47               NaT   

      order_delivered_carrier_date order_delivered_customer_date  \
1130                           NaT                           NaT   
1801                           NaT                           NaT   
1868                           NaT                           NaT   
2029                           NaT                           NaT   
2161                           NaT                           NaT   
...                            ...                           ...   
97696                          NaT                           NaT   
98415                          NaT                           NaT   
98909                          NaT                           NaT   
99283                          NaT                           NaT   
99347                          NaT                           NaT   

      order_estimated_delivery_date  
1130                     2018-09-12  
1801                     2018-10-17  
1868                     2017-04-10  
2029                     2018-09-13  
2161                     2017-05-30  
...                             ...  
97696                    2017-07-25  
98415                    2017-10-16  
98909                    2018-08-10  
99283                    2018-10-01  
99347                    2018-09-27  

[160 rows x 8 columns]

In [49]:
# fetch data that order_delivered_carrier_date have a missing value
orders_df[orders_df.order_delivered_carrier_date.isna()]


order_id                       customer_id  \
6      136cce7faa42fdb2cefd53fdc79a6098  ed0271e0b7da060a393796590e7b737a   
103    0760a852e4e9d89eb77bf631eaaf1c84  d2a79636084590b7465af8ab374a8cf5   
128    15bed8e2fec7fdbadb186b57c46c92f2  f3f0e613e0bdb9c7cee75504f0f90679   
266    8e24261a7e58791d10cb1bf9da94df5c  64a254d30eed42cd0e6c36dddb88adf0   
324    d3c8851a6651eeff2f73b0e011ac45d0  957f8e082185574de25992dc659ebbc0   
...                                 ...                               ...   
99283  3a3cddda5a7c27851bd96c3313412840  0b0d6095c5555fe083844281f6b093bb   
99313  e9e64a17afa9653aacf2616d94c005b8  b4cd0522e632e481f8eaf766a2646e86   
99347  a89abace0dcc01eeb267a9660b5ac126  2f0524a7b1b3845a1a57fcf3910c4333   
99348  a69ba794cc7deb415c3e15a0a3877e69  726f0894b5becdf952ea537d5266e543   
99415  5fabc81b6322c8443648e1b21a6fef21  32c9df889d41b0ee8309a5efb6855dcb   

      order_status order_purchase_timestamp   order_approved_at  \
6         invoiced      2017-04-11 12:22:08 2017-04-13 13:25:17   
103       invoiced      2018-08-03 17:44:42 2018-08-07 06:15:14   
128     processing      2017-09-03 14:22:03 2017-09-03 14:30:09   
266    unavailable      2017-11-16 15:09:28 2017-11-16 15:26:57   
324     processing      2016-10-05 22:44:13 2016-10-06 15:51:05   
...            ...                      ...                 ...   
99283     canceled      2018-08-31 16:13:44                 NaT   
99313   processing      2018-01-05 23:07:24 2018-01-09 07:18:05   
99347     canceled      2018-09-06 18:45:47                 NaT   
99348  unavailable      2017-08-23 16:28:04 2017-08-28 15:44:47   
99415  unavailable      2017-10-10 10:50:03 2017-10-14 18:35:57   

      order_delivered_carrier_date order_delivered_customer_date  \
6                              NaT                           NaT   
103                            NaT                           NaT   
128                            NaT                           NaT   
266                            NaT                           NaT   
324                            NaT                           NaT   
...                            ...                           ...   
99283                          NaT                           NaT   
99313                          NaT                           NaT   
99347                          NaT                           NaT   
99348                          NaT                           NaT   
99415                          NaT                           NaT   

      order_estimated_delivery_date  
6                        2017-05-09  
103                      2018-08-21  
128                      2017-10-03  
266                      2017-12-05  
324                      2016-12-09  
...                             ...  
99283                    2018-10-01  
99313                    2018-02-06  
99347                    2018-09-27  
99348                    2017-09-15  
99415                    2017-10-23  

[1783 rows x 8 columns]

In [50]:
# fetch data that order_delivered_customer_date have a missing value
orders_df[orders_df.order_delivered_customer_date.isna()]

order_id                       customer_id  \
6      136cce7faa42fdb2cefd53fdc79a6098  ed0271e0b7da060a393796590e7b737a   
44     ee64d42b8cf066f35eac1cf57de1aa85  caded193e8e47b8362864762a83db3c5   
103    0760a852e4e9d89eb77bf631eaaf1c84  d2a79636084590b7465af8ab374a8cf5   
128    15bed8e2fec7fdbadb186b57c46c92f2  f3f0e613e0bdb9c7cee75504f0f90679   
154    6942b8da583c2f9957e990d028607019  52006a9383bf149a4fb24226b173106f   
...                                 ...                               ...   
99283  3a3cddda5a7c27851bd96c3313412840  0b0d6095c5555fe083844281f6b093bb   
99313  e9e64a17afa9653aacf2616d94c005b8  b4cd0522e632e481f8eaf766a2646e86   
99347  a89abace0dcc01eeb267a9660b5ac126  2f0524a7b1b3845a1a57fcf3910c4333   
99348  a69ba794cc7deb415c3e15a0a3877e69  726f0894b5becdf952ea537d5266e543   
99415  5fabc81b6322c8443648e1b21a6fef21  32c9df889d41b0ee8309a5efb6855dcb   

      order_status order_purchase_timestamp   order_approved_at  \
6         invoiced      2017-04-11 12:22:08 2017-04-13 13:25:17   
44         shipped      2018-06-04 16:44:48 2018-06-05 04:31:18   
103       invoiced      2018-08-03 17:44:42 2018-08-07 06:15:14   
128     processing      2017-09-03 14:22:03 2017-09-03 14:30:09   
154        shipped      2018-01-10 11:33:07 2018-01-11 02:32:30   
...            ...                      ...                 ...   
99283     canceled      2018-08-31 16:13:44                 NaT   
99313   processing      2018-01-05 23:07:24 2018-01-09 07:18:05   
99347     canceled      2018-09-06 18:45:47                 NaT   
99348  unavailable      2017-08-23 16:28:04 2017-08-28 15:44:47   
99415  unavailable      2017-10-10 10:50:03 2017-10-14 18:35:57   

      order_delivered_carrier_date order_delivered_customer_date  \
6                              NaT                           NaT   
44             2018-06-05 14:32:00                           NaT   
103                            NaT                           NaT   
128                            NaT                           NaT   
154            2018-01-11 19:39:23                           NaT   
...                            ...                           ...   
99283                          NaT                           NaT   
99313                          NaT                           NaT   
99347                          NaT                           NaT   
99348                          NaT                           NaT   
99415                          NaT                           NaT   

      order_estimated_delivery_date  
6                        2017-05-09  
44                       2018-06-28  
103                      2018-08-21  
128                      2017-10-03  
154                      2018-02-07  
...                             ...  
99283                    2018-10-01  
99313                    2018-02-06  
99347                    2018-09-27  
99348                    2017-09-15  
99415                    2017-10-23  

[2965 rows x 8 columns]

In [51]:
"""
we can conclude that all the data that have a missing value have a usefull data in the order status column,
so we can not delete the data, instead of make it with None format.
"""

'\nwe can conclude that all the data that have a missing value have a usefull data in the order status column,\nso we can not delete the data, instead of make it with None format.\n'

#### 4. Cleaning Order Reviews DataFrame

In [52]:
# check datatype
order_reviews_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 99224 entries, 0 to 99223
Data columns (total 7 columns):
 #   Column                   Non-Null Count  Dtype 
---  ------                   --------------  ----- 
 0   review_id                99224 non-null  object
 1   order_id                 99224 non-null  object
 2   review_score             99224 non-null  int64 
 3   review_comment_title     11568 non-null  object
 4   review_comment_message   40977 non-null  object
 5   review_creation_date     99224 non-null  object
 6   review_answer_timestamp  99224 non-null  object
dtypes: int64(1), object(6)
memory usage: 5.3+ MB


In [53]:
# change datetime column to datetime datatype format
datetime_columns = ["review_creation_date", "review_answer_timestamp"]

for column in datetime_columns:
    order_reviews_df[column] = pd.to_datetime(order_reviews_df[column])

In [54]:
# check missing value
order_reviews_df.isna().sum()

review_id                      0
order_id                       0
review_score                   0
review_comment_title       87656
review_comment_message     58247
review_creation_date           0
review_answer_timestamp        0
dtype: int64

In [55]:
"""
here we can conclude that the column that dont have a value is an optional(nullable) value that the customer
dont have an obligation for fill the column, so we can just make it like that
"""

'\nhere we can conclude that the column that dont have a value is an optional(nullable) value that the customer\ndont have an obligation for fill the column, so we can just make it like that\n'

#### 5. Cleaning Order Items DataFrame

In [56]:
# check datatype
order_items_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 112650 entries, 0 to 112649
Data columns (total 7 columns):
 #   Column               Non-Null Count   Dtype  
---  ------               --------------   -----  
 0   order_id             112650 non-null  object 
 1   order_item_id        112650 non-null  int64  
 2   product_id           112650 non-null  object 
 3   seller_id            112650 non-null  object 
 4   shipping_limit_date  112650 non-null  object 
 5   price                112650 non-null  float64
 6   freight_value        112650 non-null  float64
dtypes: float64(2), int64(1), object(4)
memory usage: 6.0+ MB


In [57]:
# change to datetime
order_items_df["shipping_limit_date"] = pd.to_datetime(order_items_df["shipping_limit_date"])

## Exploratory Data Analysis (EDA)

### 1. Explore Customers_df Data

#### Explore Results:
- We have total 99.441 customer
- most of the customers are live in Sao Paulo City with total: 15.540 customers
- most of the customers are live in SP State with total: 41.746 customers
- total of city that customer live in is 4.119 cities
- total of state that customer live in is 27 state

In [58]:
# describe customer_df
customers_df.describe(include="all")

customer_id                customer_unique_id  \
count                              99441                             99441   
unique                             99441                             96096   
top     06b8999e2fba1a1fbc88172c00ba8bc7  8d50f5eadf50201ccdcedfb9e2ac8455   
freq                                   1                                17   
mean                                 NaN                               NaN   
std                                  NaN                               NaN   
min                                  NaN                               NaN   
25%                                  NaN                               NaN   
50%                                  NaN                               NaN   
75%                                  NaN                               NaN   
max                                  NaN                               NaN   

        customer_zip_code_prefix customer_city customer_state  
count               99441.000000         99441          99441  
unique                       NaN          4119             27  
top                          NaN     sao paulo             SP  
freq                         NaN         15540          41746  
mean                35137.474583           NaN            NaN  
std                 29797.938996           NaN            NaN  
min                  1003.000000           NaN            NaN  
25%                 11347.000000           NaN            NaN  
50%                 24416.000000           NaN            NaN  
75%                 58900.000000           NaN            NaN  
max                 99990.000000           NaN            NaN

In [59]:
# make pivot table to count customer based on city
customers_df.groupby("customer_city").customer_id.nunique().sort_values(ascending=False)

customer_city
sao paulo         15540
rio de janeiro     6882
belo horizonte     2773
brasilia           2131
curitiba           1521
                  ...  
ibiara                1
rio espera            1
rio dos indios        1
rio dos cedros        1
lagoao                1
Name: customer_id, Length: 4119, dtype: int64

In [60]:
# make pivot tabel to count customer based on state
customers_df.groupby("customer_state").customer_id.nunique().sort_values(ascending=False)

customer_state
SP    41746
RJ    12852
MG    11635
RS     5466
PR     5045
SC     3637
BA     3380
DF     2140
ES     2033
GO     2020
PE     1652
CE     1336
PA      975
MT      907
MA      747
MS      715
PB      536
PI      495
RN      485
AL      413
SE      350
TO      280
RO      253
AM      148
AC       81
AP       68
RR       46
Name: customer_id, dtype: int64

In [61]:
# count total city that customers live in
customers_df["customer_city"].nunique()

4119

In [62]:
# count total state that customers live in
customers_df["customer_state"].nunique()

27

### 2. Explore Sellers_df Data

#### Explore Results:
- we have total 3.095 seller
- most of the seller are live in Sao Paulo with total: 694 sellers
- most of the seller are live in SP state with total: 1.849 sellers
- total of city that seller live in is 611 city
- total of state that seller live in is 23 state

In [63]:
# describe seller
sellers_df.describe(include="all")

seller_id  seller_zip_code_prefix seller_city  \
count                               3095             3095.000000        3095   
unique                              3095                     NaN         611   
top     3442f8959a84dea7ee197c632cb2df15                     NaN   sao paulo   
freq                                   1                     NaN         694   
mean                                 NaN            32291.059451         NaN   
std                                  NaN            32713.453830         NaN   
min                                  NaN             1001.000000         NaN   
25%                                  NaN             7093.500000         NaN   
50%                                  NaN            14940.000000         NaN   
75%                                  NaN            64552.500000         NaN   
max                                  NaN            99730.000000         NaN   

       seller_state  
count          3095  
unique           23  
top              SP  
freq           1849  
mean            NaN  
std             NaN  
min             NaN  
25%             NaN  
50%             NaN  
75%             NaN  
max             NaN

In [64]:
# make pivot tabel to count customer based on city
sellers_df.groupby("seller_city").seller_id.nunique().sort_values(ascending=False)

seller_city
sao paulo         694
curitiba          127
rio de janeiro     96
belo horizonte     68
ribeirao preto     52
                 ... 
ivoti               1
itirapina           1
itau de minas       1
itapui              1
xaxim               1
Name: seller_id, Length: 611, dtype: int64

In [65]:
# make pivot tabel to count customer based on state
sellers_df.groupby("seller_state").seller_id.nunique().sort_values(ascending=False)

seller_state
SP    1849
PR     349
MG     244
SC     190
RJ     171
RS     129
GO      40
DF      30
ES      23
BA      19
CE      13
PE       9
PB       6
MS       5
RN       5
MT       4
RO       2
SE       2
AC       1
PI       1
AM       1
MA       1
PA       1
Name: seller_id, dtype: int64

### 3. Explore Products_df Data

#### Explore Results:
- most of the product is from category bed_bath_table with 3.029 products, and the seccond is sports_leisure category with total 2.867 products.
- average product have a 2 photos on marketplace, with min. 1 photo and max. 20 photos
- average product have 771 description length, with min. 4 words and max. 3.992 words
- average product weight is 2276.96 gram, with min. 0 gram and max. 40.425 gram

In [66]:
# describe product_df
products_df.describe(include="all")

product_id product_category_name  \
count                              32340                 32340   
unique                             32340                    73   
top     1e9e8ef04dbcff4541ed26657ea517e5       cama_mesa_banho   
freq                                   1                  3029   
mean                                 NaN                   NaN   
std                                  NaN                   NaN   
min                                  NaN                   NaN   
25%                                  NaN                   NaN   
50%                                  NaN                   NaN   
75%                                  NaN                   NaN   
max                                  NaN                   NaN   

        product_name_lenght  product_description_lenght  product_photos_qty  \
count          32340.000000                32340.000000        32340.000000   
unique                  NaN                         NaN                 NaN   
top                     NaN                         NaN                 NaN   
freq                    NaN                         NaN                 NaN   
mean              48.476592                  771.492393            2.188961   
std               10.245699                  635.124831            1.736787   
min                5.000000                    4.000000            1.000000   
25%               42.000000                  339.000000            1.000000   
50%               51.000000                  595.000000            1.000000   
75%               57.000000                  972.000000            3.000000   
max               76.000000                 3992.000000           20.000000   

        product_weight_g  product_length_cm  product_height_cm  \
count       32340.000000       32340.000000       32340.000000   
unique               NaN                NaN                NaN   
top                  NaN                NaN                NaN   
freq                 NaN                NaN                NaN   
mean         2276.956586          30.854545          16.958813   
std          4279.291845          16.955965          13.636115   
min             0.000000           7.000000           2.000000   
25%           300.000000          18.000000           8.000000   
50%           700.000000          25.000000          13.000000   
75%          1900.000000          38.000000          21.000000   
max         40425.000000         105.000000         105.000000   

        product_width_cm  
count       32340.000000  
unique               NaN  
top                  NaN  
freq                 NaN  
mean           23.208596  
std            12.078762  
min             6.000000  
25%            15.000000  
50%            20.000000  
75%            30.000000  
max           118.000000

In [67]:
"""
for more clear product category name, we will add product_category_name_english column with merge method.
merge product_df with product_category_name_translation_df
"""
products_english_df = pd.merge(
    left=products_df,
    right=product_category_name_translation_df,
    how="left",
    on="product_category_name"
)

In [68]:
# make pivot table based on product_category_name_english
products_english_df.groupby("product_category_name_english").product_id.count().sort_values(ascending=False)

product_category_name_english
bed_bath_table               3029
sports_leisure               2867
furniture_decor              2657
health_beauty                2444
housewares                   2335
                             ... 
tablets_printing_image          9
home_comfort_2                  5
fashion_childrens_clothes       5
security_and_services           2
cds_dvds_musicals               1
Name: product_id, Length: 71, dtype: int64

### 4. Explore Geolocation_df Data

#### Explore Results:
- city that have a most zip code (area) is in sao paulo city with total 3171 area
- state that have a most cities in in MG state with 1.426 cities, and then SP state with 1.048 cities
- zip code (area) that have a most location is in zip code number 38400 with 746 locations, then zipcode number 35500 have 726 locations

In [69]:
# describe dataframe
geolocation_df.describe(include="all")

geolocation_zip_code_prefix  geolocation_lat  geolocation_lng  \
count                 738332.000000    738332.000000    738332.000000   
unique                          NaN              NaN              NaN   
top                             NaN              NaN              NaN   
freq                            NaN              NaN              NaN   
mean                   38316.086800       -20.998353       -46.461098   
std                    30632.496675         5.892315         4.393705   
min                     1001.000000       -36.605374      -101.466766   
25%                    12600.000000       -23.603061       -48.867822   
50%                    29144.000000       -22.873588       -46.647278   
75%                    65950.000000       -19.923336       -43.836974   
max                    99990.000000        45.065933       121.105394   

       geolocation_city geolocation_state  
count            738332            738332  
unique             8011                27  
top           sao paulo                SP  
freq              79927            285976  
mean                NaN               NaN  
std                 NaN               NaN  
min                 NaN               NaN  
25%                 NaN               NaN  
50%                 NaN               NaN  
75%                 NaN               NaN  
max                 NaN               NaN

In [70]:
# make pivot tb. group by city to find how many zip_code per city
geolocation_df.groupby("geolocation_city").geolocation_zip_code_prefix.nunique().sort_values(ascending=False)

geolocation_city
sao paulo               3171
são paulo               3013
brasilia                 496
brasília                 406
rio de janeiro           404
                        ... 
guarei                     1
guardinha                  1
guarda-mor                 1
guarda dos ferreiros       1
óleo                       1
Name: geolocation_zip_code_prefix, Length: 8011, dtype: int64

In [76]:
# make a pivot table group by state, how many city in one state
geolocation_df.groupby("geolocation_state").geolocation_city.nunique().sort_values(ascending=False)

geolocation_state
MG    1426
SP    1048
RS     691
BA     652
PR     651
SC     420
GO     384
MA     299
PI     278
PE     267
CE     260
PB     254
RJ     245
PA     219
RN     214
MT     213
TO     173
ES     160
MS     133
AL     130
SE      96
RO      83
AM      74
AC      34
DF      28
AP      17
RR      14
Name: geolocation_city, dtype: int64

In [79]:
# make a pivot table to see each zip code (area) have how many location
geolocation_df.groupby("geolocation_zip_code_prefix").agg({
    "geolocation_lat": "nunique",
    "geolocation_lng": "nunique"
}).sort_values(by="geolocation_lat", ascending=False)

geolocation_lat  geolocation_lng
geolocation_zip_code_prefix                                  
38400                                    746              745
35500                                    726              726
11680                                    726              726
11740                                    666              666
36400                                    627              627
...                                      ...              ...
11620                                      1                1
11625                                      1                1
35150                                      1                1
91509                                      1                1
25928                                      1                1

[19015 rows x 2 columns]

### 5. Explore Order_df Data

#### Explore Results:

- we have total 99441 order
- each customer just do 1 order
- most of the order have done with delivered status in total 96.478 orders done
- the order that have been canceled have a total 625 orders
- the order that is still in process have a total 301 orders
- the order that is unavailable have a total 609 orders
- the order that is still in shipment process is 1.107 orders
- seller performance to process approvable after customer purchased is 

In [81]:
# describe order_df
orders_df.describe(include="all")

order_id                       customer_id  \
count                              99441                             99441   
unique                             99441                             99441   
top     e481f51cbdc54678b7cc49136f2d6af7  9ef432eb6251297304e76186b10a928d   
freq                                   1                                 1   
mean                                 NaN                               NaN   
min                                  NaN                               NaN   
25%                                  NaN                               NaN   
50%                                  NaN                               NaN   
75%                                  NaN                               NaN   
max                                  NaN                               NaN   

       order_status       order_purchase_timestamp  \
count         99441                          99441   
unique            8                            NaN   
top       delivered                            NaN   
freq          96478                            NaN   
mean            NaN  2017-12-31 08:43:12.776581120   
min             NaN            2016-09-04 21:15:19   
25%             NaN            2017-09-12 14:46:19   
50%             NaN            2018-01-18 23:04:36   
75%             NaN            2018-05-04 15:42:16   
max             NaN            2018-10-17 17:30:18   

                    order_approved_at   order_delivered_carrier_date  \
count                           99281                          97658   
unique                            NaN                            NaN   
top                               NaN                            NaN   
freq                              NaN                            NaN   
mean    2017-12-31 18:35:24.098800128  2018-01-04 21:49:48.138278656   
min               2016-09-15 12:16:38            2016-10-08 10:34:01   
25%               2017-09-12 23:24:16  2017-09-15 22:28:50.249999872   
50%               2018-01-19 11:36:13            2018-01-24 16:10:58   
75%               2018-05-04 20:35:10            2018-05-08 13:37:45   
max               2018-09-03 17:40:06            2018-09-11 19:48:28   

        order_delivered_customer_date  order_estimated_delivery_date  
count                           96476                          99441  
unique                            NaN                            NaN  
top                               NaN                            NaN  
freq                              NaN                            NaN  
mean    2018-01-14 12:09:19.035542272  2018-01-24 03:08:37.730111232  
min               2016-10-11 13:46:32            2016-09-30 00:00:00  
25%     2017-09-25 22:07:22.249999872            2017-10-03 00:00:00  
50%        2018-02-02 19:28:10.500000            2018-02-15 00:00:00  
75%     2018-05-15 22:48:52.249999872            2018-05-25 00:00:00  
max               2018-10-17 13:22:46            2018-11-12 00:00:00

In [82]:
# make pivot table based on order_status, to count how many each each order status
orders_df.groupby("order_status").order_id.nunique().sort_values(ascending=False)

order_status
delivered      96478
shipped         1107
canceled         625
unavailable      609
invoiced         314
processing       301
created            5
approved           2
Name: order_id, dtype: int64

In [83]:
# make a pivot tabel grop by order_id to see how many order each customer
orders_df.groupby("order_id").customer_id.nunique().sort_values(ascending=False)

order_id
00010242fe8c5a6d1ba2dd792cb16214    1
ab291c01a71273689c1e6ea624b083dc    1
ab344aaf7040ad7bcee8c7ae7db9e87e    1
ab30f8d4ebb7d33d1dcf54d420ea42cd    1
ab30f399d360ddf62897051d813f887d    1
                                   ..
55539e6edbbf2276dc9e2eab8af0fa8d    1
55538582ef35a95e5ed4047c6bf04695    1
5553543e89ef381e4ff08f6b91b94690    1
55534471fcbe2dd1436234e7cfa4de5a    1
fffe41c64501cc87c801fd61db3f6244    1
Name: customer_id, Length: 99441, dtype: int64

In [119]:
# seller performance to process approvable after customer purchased
# count days range between order purchased and approve
range_order_purchased_to_approved = orders_df["order_approved_at"] - orders_df["order_purchase_timestamp"]
range_order_purchased_to_approved = range_order_purchased_to_approved.apply(lambda x: x.total_seconds()//86400)

In [122]:
# time customer must wait from first do the purchase until come to the customer home
# count days range between order purchased and order delivered
range_order_purchased_to_delivered = orders_df["order_delivered_customer_date"] - orders_df["order_purchase_timestamp"]
range_order_purchased_to_delivered = range_order_purchased_to_delivered.apply(lambda x: x.total_seconds()//86400)

In [128]:
# carrier performance
# count days range between order delivered to carrier and order delivered to customer
range_order_delivered_to_carrier_and_delivered_to_customer = orders_df["order_delivered_customer_date"] - orders_df["order_delivered_carrier_date"]
range_order_delivered_to_carrier_and_delivered_to_customer = range_order_delivered_to_carrier_and_delivered_to_customer.apply(lambda x: x.total_seconds()//86400)

In [124]:
# order delivery accuracy time
# count days range between order delivered to customer and order estimated delivery
# range_order_delivered_customer_and_estimated_delivery_date = orders_df["order_estimated_delivery_date"] - orders_df["order_delivered_customer_date"]
# range_order_delivered_customer_and_estimated_delivery_date.apply(lambda x: x.total_seconds()//86400).sort_values(ascending=False)

In [130]:
# make new order_performance data frame to track performance
d = {
    "order_id": orders_df["order_id"].tolist(),
    "order_purchased_to_approved_time": range_order_purchased_to_approved.tolist(),
    "order_purchased_to_delivered_time": range_order_purchased_to_delivered.tolist(),
    "order_from_carrier_to_delivered_time": range_order_delivered_to_carrier_and_delivered_to_customer.tolist()
}
order_performance_df = pd.DataFrame(data=d)
order_performance_df.describe()

order_purchased_to_approved_time  order_purchased_to_delivered_time  \
count                      99281.000000                       96476.000000   
mean                           0.269800                          12.094086   
std                            0.986202                           9.551746   
min                            0.000000                           0.000000   
25%                            0.000000                           6.000000   
50%                            0.000000                          10.000000   
75%                            0.000000                          15.000000   
max                          187.000000                         209.000000   

       order_from_carrier_to_delivered_time  
count                          96475.000000  
mean                               8.878310  
std                                8.746088  
min                              -17.000000  
25%                                4.000000  
50%                                7.000000  
75%                               12.000000  
max                              205.000000

## Visualization & Explanatory Analysis

### Pertanyaan 1: 

### Pertanyaan 2: 

## Conclusion

- Conclution pertanyaan 1
- conclution pertanyaan 2